README.md
# QUnit.m

[![Version](http://cocoapod-badges.herokuapp.com/v/QUnit.m/badge.png)](http://cocoadocs.org/docsets/QUnit.m)
[![Platform](http://cocoapod-badges.herokuapp.com/p/QUnit.m/badge.png)](http://cocoadocs.org/docsets/QUnit.m)

QUnit.m brings [QUnit.js](http://qunitjs.com/) syntax to SenTestingKit for testing Objective-C projects. The is useful if you are trying to port JavaScript libraries to Objective-C or want a more familiar testing syntax.

I wrote this to help me port Underscore.js to Objective-C in [_.m](https://github.com/kmalakoff/_.m) and to make Objective-C a little more script in [Subjective-Script](https://github.com/kmalakoff/SubjectiveScript.m) so you will find some good examples in those project's tests.

**Note:** the library is still in its early days so please submit pull requests with corresponding tests to fix any issues and to bring the feature set up to the level of QUnit.js!

Examples (some from _.m and Subjective-Script)
------------

equal

```
equal(AI(1,2,3).reverse().join(@","), @"3,2,1", @"reverse works");
```

strictEqual

```
strictEqual(_.chain(array).find(^B(N* n) { return n.I > 2; }).N, N.I(3), @"should return first found `value`");
```

deepEqual

```
_WrappedBlock wrapped = _.wrap(noop, ^A*(_WrappedBlock fn, id arg1, ...){ ARGS_AO(arguments, arg1); return A.newNSO(fn).concat(arguments); });
A* ret     = wrapped(AO(@"whats", @"your"), @"vector", @"victor", nil);
deepEqual(ret, AO(noop, AO(@"whats", @"your"), @"vector", @"victor"), @"deep wrapping");
```

notEqual

```
notEqual(AI(1,2,3).reverse().join(@","), @"1,2,3", @"reverse works");
```

ok

```
ok(OKV({@"int", N.I(1)}, {@"float", N.F(2.0)}).hasOwnProperty(@"int"), @"has an int");
```

raises

```
raises(^{ _.reduce(A.new, ^(N* memo, N* num, ...){ return memo; }, 0); }, @"TypeError", @"throws an error for empty arrays with no initial value");
```

asyncTest

```
asyncTest(^{
  __block BOOL called1 = false;
  SSTimeout* timeout1 = SS.setTimeout(^{ called1 = true; }, NSEC_PER_SEC*1);
  SS.clearTimeout(timeout1);

  __block BOOL called2 = false;
  SS.setTimeout(^{ called2 = true; start(); }, NSEC_PER_SEC*2);
});
```

## Usage

To run the example project; clone the repo, and run `pod install` from the Project directory first.

## Requirements

## Installation

QUnit.m is available through [CocoaPods](http://cocoapods.org), to install
it simply add the following line to your Podfile:

    pod "QUnit.m"

## Author

Kevin Malakoff, kmalakoff@gmail.com

## License

QUnit.m is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# QUnit.m CHANGELOG

## 0.1.0

Initial release.


Tests/QUnitTestsTests.m
//
//  QUnitTestsTests.m
//  QUnitTests
//
//  Created by Kevin Malakoff on 7/29/12.
//  Copyright (c) 2012 Kevin Malakoff. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person
//  obtaining a copy of this software and associated documentation
//  files (the "Software"), to deal in the Software without
//  restriction, including without limitation the rights to use,
//  copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the
//  Software is furnished to do so, subject to the following
//  conditions:
//
//  The above copyright notice and this permission notice shall be
//  included in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
//  OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
//  WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
//  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
//  OTHER DEALINGS IN THE SOFTWARE.
//

#import "QUnitTestsTests.h"
#import "QUnit.h"

@implementation QUnitTestsTests

- (void)test_version
{
  equal(QUnit.VERSION, @"0.1.1", @"expected version");
}

- (void)testObjectiveCTypes
{
  equal(@"hello world", @"hello world", @"equal: constant strings work")
  notEqual(@"hello world", @"hello world!", @"notEqual: constant strings work")

  equal([NSNumber numberWithInteger:1], [NSNumber numberWithInteger:1], @"equal: NSNumbers work")
  notEqual([NSNumber numberWithInteger:1], [NSNumber numberWithInteger:-1], @"notEqual: NSNumbers work")
}

- (void)testPrimitiveTypes
{
  equal(1, 1, @"equal: 1 == 1")
  notEqual(1, -1, @"equal: 1 != -1")

  equal(true, true, @"equal: true == true")
  equal(false, false, @"equal: false != false")
  notEqual(false, true, @"equal: false != true")
  notEqual(true, false, @"equal: true != false")

  equal(YES, YES, @"equal: YES == YES")
  equal(NO, NO, @"equal: NO != NO")
  notEqual(NO, YES, @"equal: NO != YES")
  notEqual(YES, NO, @"equal: YES != NO")

  equal(1.2f, 1.2f, @"equal: 1.2 == 1.2")
  notEqual(1.2f, 3.14159, @"equal: 1.2 != 3.14159")
}

@end


QUnitTests/QUnitTestsTests.h
//
//  QUnitTestsTests.h
//  QUnitTests
//
//  Created by Kevin Malakoff on 7/29/12.
//  Copyright (c) 2012 Kevin Malakoff. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person
//  obtaining a copy of this software and associated documentation
//  files (the "Software"), to deal in the Software without
//  restriction, including without limitation the rights to use,
//  copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the
//  Software is furnished to do so, subject to the following
//  conditions:
//
//  The above copyright notice and this permission notice shall be
//  included in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
//  OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
//  WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
//  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
//  OTHER DEALINGS IN THE SOFTWARE.
//

#import <SenTestingKit/SenTestingKit.h>

@interface QUnitTestsTests : SenTestCase

@end


QUnitTests/QUnitTestsTests.m
//
//  QUnitTestsTests.m
//  QUnitTests
//
//  Created by Kevin Malakoff on 7/29/12.
//  Copyright (c) 2012 Kevin Malakoff. All rights reserved.
//
//  Permission is hereby granted, free of charge, to any person
//  obtaining a copy of this software and associated documentation
//  files (the "Software"), to deal in the Software without
//  restriction, including without limitation the rights to use,
//  copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the
//  Software is furnished to do so, subject to the following
//  conditions:
//
//  The above copyright notice and this permission notice shall be
//  included in all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
//  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
//  OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
//  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
//  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
//  WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
//  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
//  OTHER DEALINGS IN THE SOFTWARE.
//

#import "QUnitTestsTests.h"
#import "QUnit.h"

@implementation QUnitTestsTests

- (void)test_version
{
  equal(QUnit.VERSION, @"0.1.0", @"expected version");
}

- (void)testObjectiveCTypes
{
  equal(@"hello world", @"hello world", @"equal: constant strings work")
  notEqual(@"hello world", @"hello world!", @"notEqual: constant strings work")

  equal([NSNumber numberWithInteger:1], [NSNumber numberWithInteger:1], @"equal: NSNumbers work")
  notEqual([NSNumber numberWithInteger:1], [NSNumber numberWithInteger:-1], @"notEqual: NSNumbers work")
}

- (void)testPrimitiveTypes
{
  equal(1, 1, @"equal: 1 == 1")
  notEqual(1, -1, @"equal: 1 != -1")

  equal(true, true, @"equal: true == true")
  equal(false, false, @"equal: false != false")
  notEqual(false, true, @"equal: false != true")
  notEqual(true, false, @"equal: true != false")

  equal(YES, YES, @"equal: YES == YES")
  equal(NO, NO, @"equal: NO != NO")
  notEqual(NO, YES, @"equal: NO != YES")
  notEqual(YES, NO, @"equal: YES != NO")

  equal(1.2f, 1.2f, @"equal: 1.2 == 1.2")
  notEqual(1.2f, 3.14159, @"equal: 1.2 != 3.14159")
}

@end
